diff --git a/src/backend/Headless.cpp b/src/backend/Headless.cpp
index 5d735d6..140d70c 100644
--- a/src/backend/Headless.cpp
+++ b/src/backend/Headless.cpp
@@ -17,13 +17,6 @@ using namespace Hyprutils::Math;
 Aquamarine::CHeadlessOutput::CHeadlessOutput(const std::string& name_, Hyprutils::Memory::CWeakPointer<CHeadlessBackend> backend_) : backend(backend_) {
     name = name_;
 
-    framecb = makeShared<std::function<void()>>([this]() {
-        frameScheduled = false;
-        // not sure about removing this since framecb might be called outside of scheduleFrames no?!
-        lastFrame      = std::chrono::steady_clock::now();
-        events.frame.emit();
-    });
-
     lastFrame = std::chrono::steady_clock::now();
 }
 
@@ -61,6 +54,18 @@ void Aquamarine::CHeadlessOutput::scheduleFrame(const scheduleFrameReason reason
 
     needsFrame = true;
 
+    if (!framecb) {
+        framecb = makeShared<std::function<void()>>([this, weak = self]() {
+            if (!weak)
+                return;
+
+            frameScheduled = false;
+            // not sure about removing this since framecb might be called outside of scheduleFrames no?!
+            lastFrame = std::chrono::steady_clock::now();
+            events.frame.emit();
+        });
+    }
+
     if (frameScheduled)
         return;
 
@@ -89,7 +94,7 @@ void Aquamarine::CHeadlessOutput::scheduleFrame(const scheduleFrameReason reason
 
         if (framecb && *framecb)
             (*framecb)();
-        
+
         lastFrame = NEXT_FRAME_TIME;
     });
 }
@@ -204,13 +209,9 @@ void Aquamarine::CHeadlessBackend::dispatchTimers() {
         return;
     }
 
-    auto it = std::stable_partition(timers.timers.begin(), timers.timers.end(), 
-        [](CTimer& t) {
-            return !t.expired();
-        });
-    
-    std::vector<CTimer> toFire(std::make_move_iterator(it), 
-                               std::make_move_iterator(timers.timers.end()));
+    auto                it = std::stable_partition(timers.timers.begin(), timers.timers.end(), [](CTimer& t) { return !t.expired(); });
+
+    std::vector<CTimer> toFire(std::make_move_iterator(it), std::make_move_iterator(timers.timers.end()));
     timers.timers.erase(it, timers.timers.end());
 
     for (auto& timer : toFire) {
